### PR TITLE
Currency: Add new forex provider exchangerate.host

### PR DESCRIPTION
**Pull request by thrasher-**
# PR Description

Adds a new default forex provider ExchangeRateHost.io due to ExchangeRateAPI disallowing APIkey free requests.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by shazbert — CHANGES_REQUESTED** [review]
A couple of tests were failing for exchange rate host due to a time issue. Other than that. LGTM.

File: config/config.go [hunk 1]
```diff
@@ -1086,11 +1086,12 @@ func (c *Config) CheckCurrencyConfigValues() error {
 	count := 0
 	for i := range c.Currency.ForexProviders {
 		if c.Currency.ForexProviders[i].Enabled {
-			if c.Currency.ForexProviders[i].Name == "CurrencyConverter" &&
+			if (c.Currency.ForexProviders[i].Name == "CurrencyConverter" || c.Currency.ForexProviders[i].Name == "ExchangeRates") &&
 				c.Currency.ForexProviders[i].PrimaryProvider &&
 				(c.Currency.ForexProviders[i].APIKey == "" ||
 					c.Currency.ForexProviders[i].APIKey == DefaultUnsetAPIKey) {
-				log.Warnln(log.Global, "CurrencyConverter forex provider no longer supports unset API key requests. Switching to ExchangeRates FX provider..")
+				log.Warnf(log.Global, "%s forex provider no longer supports unset API key requests. Switching to %s FX provider..",
```
Comment (gloriousCode) [inline config/config.go:1093]: Can't highlight it specifically, but my config upgrade experience does not go through this path.
Instead I receive:
```
ExchangeRatesHost forex provider not found, adding to config..
ExchangeRates enabled forex provider API key not set. Please set this in your config.json file
Using ExchangeRatesAPI for default forex provider.
```
And yet `ExchangeRatesHost` is the one enabled.

File: currency/forexprovider/exchangeratesapi.io/exchangeratesapi.go [hunk 2]
```diff
@@ -47,34 +63,45 @@ func cleanCurrencies(baseCurrency, symbols string) string {
...
 // GetLatestRates returns a map of forex rates based on the supplied params
 // baseCurrency - USD	[optional] The base currency to use for forex rates, defaults to EUR
 // symbols - AUD,USD	[optional] The symbols to query the forex rates for, default is
 // all supported currencies
-func (e *ExchangeRates) GetLatestRates(baseCurrency, symbols string) (Rates, error) {
+func (e *ExchangeRates) GetLatestRates(baseCurrency, symbols string) (*Rates, error) {
 	vals := url.Values{}
-
-	if len(baseCurrency) > 0 {
+	if len(baseCurrency) > 0 && e.APIKeyLvl <= apiKeyFree && !reflect.DeepEqual("EUR", baseCurrency) {
```
Comment (gloriousCode) [inline currency/forexprovider/exchangeratesapi.io/exchangeratesapi.go:92]: I'm unsure why you've used reflect.DeepEqual for a string comparison. Perhaps strings.Equalfold? Or even `!=` if you're not checking case
``` 
BenchmarkDeepEqual-8   	58.34 ns/op
BenchmarkEqualFold-8   	5.040 ns/op
BenchmarkEquals-8   	0.2716 ns/op
```

File: currency/forexprovider/exchangeratesapi.io/exchangeratesapi.go [hunk 3]
```diff
@@ -83,20 +110,48 @@ func (e *ExchangeRates) GetLatestRates(baseCurrency, symbols string) (Rates, err
...
-func (e *ExchangeRates) GetHistoricalRates(date, baseCurrency string, symbols []string) (HistoricalRates, error) {
+func (e *ExchangeRates) GetHistoricalRates(date time.Time, baseCurrency string, symbols []string) (*HistoricalRates, error) {
+	if date.IsZero() {
+		return nil, errors.New("a date must be specified")
+	}
+
 	var resp HistoricalRates
 	v := url.Values{}
 
+	if len(baseCurrency) > 0 && e.APIKeyLvl <= apiKeyFree && !reflect.DeepEqual("EUR", baseCurrency) {
```
Comment (gloriousCode) [inline currency/forexprovider/exchangeratesapi.io/exchangeratesapi.go:121]: Same here wrt deepequal

File: currency/forexprovider/exchangerate.host/exchangerate.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+		targetSource = source
+	}
+	v.Set("source", targetSource)
+
+	var l LatestRates
+	return &l, e.SendHTTPRequest("latest", v, &l)
+}
+
+// ConvertCurrency converts a currency based on the supplied params
+func (e *ExchangeRateHost) ConvertCurrency(from, to string, date time.Time, baseCurrency, symbols string, amount float64, places int64, source string) (*ConvertCurrency, error) {
```
Comment (gloriousCode) [inline currency/forexprovider/exchangerate.host/exchangerate.go:72]: Incredibly ~petty~ minor, but could just make it
`ConvertCurrency(from, to, baseCurrency, symbols, source string, date time.Time, amount float64, places int64)`
to minimise `string`. Can resolve if for variable flow reasons

**Review by gloriousCode — CHANGES_REQUESTED** [review]
I'm very pleased that we can get some working tests again! Thank you for this and the additional protections added. Just some minor things.

File: currency/forexprovider/exchangerate.host/exchangerate.go (hunk 4 above)
Comment (thrasher-) [reply currency/forexprovider/exchangerate.host/exchangerate.go]: Done, this was just to mirror the API ordering at the time to ensure everything was supported correctly (minus callback and format)

**Review by gloriousCode — COMMENTED** [review]
More a question than a request to change. Each of the other FX providers have a bot setting:
```
	// Forex settings
	EnableCurrencyConverter bool
	EnableCurrencyLayer     bool
	EnableFixer             bool
	EnableOpenExchangeRates bool
```
Which sets up a storage updater in `engine.go` line 417 with `err = currency.RunStorageUpdater(currency.BotOverrides{....})`. This storage.go RunUpdater does not have a `case` for exchangeratehost.
These are used to override config file settings, would you wish to add this for the new exchangerate.host?

**Comment by thrasher-** [comment]
> More a question than a request to change. Each of the other FX providers have a bot setting:
> 
> ```
> 	// Forex settings
> 	EnableCurrencyConverter bool
> 	EnableCurrencyLayer     bool
> 	EnableFixer             bool
> 	EnableOpenExchangeRates bool
> ```
> 
> Which sets up a storage updater in `engine.go` line 417 with `err = currency.RunStorageUpdater(currency.BotOverrides{....})`. This storage.go RunUpdater does not have a `case` for exchangeratehost.
> These are used to override config file settings, would you wish to add this for the new exchangerate.host?

Noticed that missing on my last passthrough, added!

**Review by MadCozBadd — APPROVED** [review]
TACK!!!! Am kind of new to this section of the code so just tested the rates provided with GCT and with tests and compared them to live rates and all looks good!

**Review by shazbert — APPROVED** [review]
tACK 👍

File: engine/engine.go [hunk 5]
```diff
@@ -563,7 +565,8 @@ func (bot *Engine) Stop() {
 		bot.Settings.EnableCurrencyConverter ||
 		bot.Settings.EnableCurrencyLayer ||
 		bot.Settings.EnableFixer ||
-		bot.Settings.EnableOpenExchangeRates {
+		bot.Settings.EnableOpenExchangeRates ||
+		bot.Settings.EnableExchangeRateHost {
```
Comment (gloriousCode) [inline engine/engine.go:569]: Please add this `|| bot.Settings.EnableExchangeRateHost` to start as well

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! One more thing

**Comment by thrasher-** [comment]
NOTE: Engine has no linkage to forex provider override settings. This will be fixed in a separate PR.

**Review by gloriousCode — APPROVED** [review]
Thank you for those changes! Thank you for fixing long running errors and making improvements. tACK!

**Review by shazbert — APPROVED** [review]
Changes ACK.

**Review by MadCozBadd — APPROVED** [review]
💯

**Pull request merged**